core/src/main/java/nu/yona/server/rest/TestController.java

```diff
@@ -153,31 +154,52 @@ private FirebaseMessageResourceAssembler createFirebaseMessageResourceAssembler(
 		private final String title;
 		private final String body;
 		private final Map<String, String> data;
+		private final String appOs;
+		private final int appVersionCode;
+		private final String appVersionName;
 
-		public FirebaseMessageDto(String title, String body, Map<String, String> data)
+		public FirebaseMessageDto(String title, String body, Map<String, String> data, String appOs, int appVersionCode,
+				String appVersionName)
 		{
 			this.title = title;
 			this.body = body;
 			this.data = data;
+			this.appOs = appOs;
+			this.appVersionCode = appVersionCode;
+			this.appVersionName = appVersionName;
 		}
 
 		@SuppressWarnings("unchecked")
-		public static FirebaseMessageDto createInstance(Message message)
+		public static FirebaseMessageDto createInstance(MessageData messageData)
 		{
 			try
 			{
-				Notification notification = (Notification) notificationField.get(message);
-				Map<String, String> data = (Map<String, String>) dataField.get(message);
+				Notification notification = (Notification) notificationField.get(messageData.firebaseMessage);
+				Map<String, String> data = (Map<String, String>) dataField.get(messageData.firebaseMessage);
 				String title = (String) titleField.get(notification);
 				String body = (String) bodyField.get(notification);
-				return new FirebaseMessageDto(title, body, data);
+				String appOs = getStringValueFromMdc(messageData, Constants.APP_OS_MDC_KEY).orElse(null);
```

Comment:
Isn't it a bit strange to get this from MDC while it is present in `HeadersHolder`?

Reply:
The test controller is here to verify the firebase messaging and related logging. For that reason, it takes it from the place LogBack would take it from as well: the MDC. Makes sense?

Reply:
Wouldn't somebody understand it better when it is plainly got from the `HeadersHolder`? `MDC` is only for logging, while this method is about getting the app information, not the logging.

Reply:
No. The headers holder is just a container/holder to hold the headers that are passed through in HTTP requests. We use some information that's communicated through headers as "mapped diagnostic context" in logging. The controller enables a test to see that the right information made it into this diagnostic context. It did so using the pass through headers solution, but the integration test verifies the result not the means.